File: articles/gender-development-and-sexual-development/r/model_gender_sexual_development.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "gender_sexual_development_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_gender_sexual_development_panel.py first.")

panel <- read.csv(data_path)

model <- lmer(
  adjustment_score ~ pubertal_progress + current_family_support +
    current_recognition + current_consent_knowledge +
    current_connectedness + school_climate +
    health_education_quality + anti_harassment_support +
    current_stigma + chronic_stigma +
    current_stigma:protective_context +
    (1 + pubertal_progress | school_id/id),
  data = panel
)

capture.output(
  summary(model),
  file = file.path(outputs_dir, "r_gender_sexual_development_model_summary.txt")
)

trajectory <- panel |>
  group_by(time, chronic_stigma) |>
  summarize(
    average_adjustment = mean(adjustment_score),
    average_protective_context = mean(protective_context),
    average_stigma = mean(current_stigma),
    standard_error = sd(adjustment_score) / sqrt(n()),
    .groups = "drop"
  ) |>
  mutate(
    lower = average_adjustment - 1.96 * standard_error,
    upper = average_adjustment + 1.96 * standard_error,
    group_label = ifelse(chronic_stigma == 1, "Higher stigma risk", "Lower stigma risk")
  )

write.csv(trajectory, file.path(outputs_dir, "r_stigma_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_stigma_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_adjustment, linetype = group_label)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper, group = group_label), alpha = 0.12) +
    labs(title = "Synthetic Gender Development, Sexual Development, and Adjustment", x = "Time", y = "Average adjustment", linetype = "Group") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

support_summary <- panel |>
  group_by(time) |>
  summarize(
    average_family_support = mean(current_family_support),
    average_recognition = mean(current_recognition),
    average_consent_knowledge = mean(current_consent_knowledge),
    average_connectedness = mean(current_connectedness),
    average_stigma = mean(current_stigma),
    average_protective_context = mean(protective_context),
    average_adjustment = mean(adjustment_score),
    .groups = "drop"
  )

write.csv(support_summary, file.path(outputs_dir, "r_protective_context_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_protective_context_trajectory.png"),
  plot = ggplot(support_summary, aes(x = time)) +
    geom_line(aes(y = average_family_support, linetype = "family support"), linewidth = 1) +
    geom_line(aes(y = average_recognition, linetype = "recognition"), linewidth = 1) +
    geom_line(aes(y = average_consent_knowledge, linetype = "consent knowledge"), linewidth = 1) +
    geom_line(aes(y = average_connectedness, linetype = "connectedness"), linewidth = 1) +
    geom_line(aes(y = average_stigma, linetype = "stigma"), linewidth = 1) +
    labs(title = "Synthetic Protective Context and Stigma Over Time", x = "Time", y = "Average index", linetype = "Measure") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R gender/sexual development outputs.")

File: articles/gender-development-and-sexual-development/python/model_gender_sexual_development.py
#!/usr/bin/env python3
"""Model synthetic gender development, sexual development, support, and stigma."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "gender_sexual_development_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_gender_sexual_development_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("id")["adjustment_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    adjustment_model = smf.ols(
        """
        adjustment_score ~ lag_score + pubertal_progress +
        current_family_support + current_recognition +
        current_consent_knowledge + current_connectedness +
        school_climate + health_education_quality +
        anti_harassment_support + current_stigma + chronic_stigma +
        current_stigma:protective_context
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    protective_model = smf.ols(
        """
        adjustment_score ~ pubertal_progress + protective_context +
        current_stigma + chronic_stigma + protective_context:current_stigma +
        C(development_profile)
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_gender_sexual_development_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC GENDER/SEXUAL DEVELOPMENT ADJUSTMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(adjustment_model.summary().as_text())
        f.write("\n\nPROTECTIVE CONTEXT AND STIGMA MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(protective_model.summary().as_text())

    trajectory = panel.groupby(["time", "chronic_stigma"], as_index=False).agg(
        average_adjustment=("adjustment_score", "mean"),
        average_protective_context=("protective_context", "mean"),
        average_stigma=("current_stigma", "mean"),
        standard_error=("adjustment_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["group_label"] = trajectory["chronic_stigma"].map({
        0: "Lower stigma risk",
        1: "Higher stigma risk",
    })
    trajectory["lower"] = trajectory["average_adjustment"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_adjustment"] + 1.96 * trajectory["standard_error"]
    trajectory.to_csv(OUTPUTS_DIR / "python_stigma_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in trajectory.groupby("group_label"):
        plt.plot(group["time"], group["average_adjustment"], marker="o", label=str(group_name))
    plt.xlabel("Time")
    plt.ylabel("Average adjustment score")
    plt.title("Synthetic Gender Development, Sexual Development, and Adjustment")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_stigma_trajectory.png", dpi=160)
    plt.close()

    protective = panel.groupby("time", as_index=False).agg(
        average_family_support=("current_family_support", "mean"),
        average_recognition=("current_recognition", "mean"),
        average_consent_knowledge=("current_consent_knowledge", "mean"),
        average_connectedness=("current_connectedness", "mean"),
        average_stigma=("current_stigma", "mean"),
        average_protective_context=("protective_context", "mean"),
        average_adjustment=("adjustment_score", "mean"),
    )
    protective.to_csv(OUTPUTS_DIR / "python_protective_context_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(protective["time"], protective["average_family_support"], marker="o", label="family support")
    plt.plot(protective["time"], protective["average_recognition"], marker="o", label="recognition")
    plt.plot(protective["time"], protective["average_consent_knowledge"], marker="o", label="consent knowledge")
    plt.plot(protective["time"], protective["average_connectedness"], marker="o", label="connectedness")
    plt.plot(protective["time"], protective["average_stigma"], marker="o", label="stigma")
    plt.xlabel("Time")
    plt.ylabel("Average index")
    plt.title("Synthetic Protective Context and Stigma Over Time")
    plt.legend(fontsize=8)
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_protective_context_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("development_profile", as_index=False).agg(
        adolescents=("id", "nunique"),
        average_adjustment=("adjustment_score", "mean"),
        average_protective_context=("protective_context", "mean"),
        average_stigma=("current_stigma", "mean"),
        average_family_support=("current_family_support", "mean"),
        average_recognition=("current_recognition", "mean"),
        average_consent_knowledge=("current_consent_knowledge", "mean"),
        average_connectedness=("current_connectedness", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_adolescent_development_profiles.csv", index=False)

    print("Wrote Python gender/sexual development model outputs.")


if __name__ == "__main__":
    main()
